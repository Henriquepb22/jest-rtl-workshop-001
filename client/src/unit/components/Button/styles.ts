import { Colors, Sizes } from '.'

const baseButtonClasses = [
  'rounded',
  'w-full',
  'transition',
  'hover:opacity-80',
  'disabled:opacity-70',
  'disabled:cursor-not-allowed'
]

const buttonSizesClassesVariants = {
  sm: ['px-2', 'py-1', 'text-sm'],
  md: ['px-4', 'py-2', 'text-base'],
  lg: ['px-6', 'py-3', 'text-lg']
}

const buttonColorClassesVariants = {
  primary: ['bg-pink-600', 'text-white'],
  secondary: ['bg-gray-300', 'text-black']
}

const buttonClasses = (color: Colors, size: Sizes) =>
  [
    ...buttonSizesClassesVariants[size],
    ...buttonColorClassesVariants[color],
    ...baseButtonClasses
  ].join(' ')

export { buttonClasses }
