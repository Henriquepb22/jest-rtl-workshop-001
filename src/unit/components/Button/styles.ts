import { Sizes } from '.'

const buttonClassesVariants = {
  sm: ['w-8', 'p-1'],
  md: ['w-10', 'p-2'],
  lg: ['w-14', 'p-3']
}

const buttonClasses = (size: Sizes) => buttonClassesVariants[size].join(' ')

export { buttonClasses }
