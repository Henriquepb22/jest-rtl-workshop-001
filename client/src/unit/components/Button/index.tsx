import { ButtonHTMLAttributes } from 'react'
import { buttonClasses } from './styles'

export type Sizes = 'sm' | 'md' | 'lg'
export type Colors = 'primary' | 'secondary'

type ButtonProps = {
  size?: Sizes
  color?: Colors
} & ButtonHTMLAttributes<HTMLButtonElement>

const Button = ({
  size = 'md',
  color = 'primary',
  children,
  ...props
}: ButtonProps) => {
  return (
    <button className={buttonClasses(color, size)} {...props}>
      {!!children && <span>{children}</span>}
    </button>
  )
}

export default Button
