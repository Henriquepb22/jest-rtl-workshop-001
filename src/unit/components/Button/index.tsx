import { buttonClasses } from './styles'

export type Sizes = 'sm' | 'md' | 'lg'

type ButtonProps = {
  onClick: () => void
  size?: Sizes
  children: React.ReactNode
}

const Button = ({ onClick, size = 'md', children }: ButtonProps) => {
  return (
    <button className={buttonClasses(size!)} onClick={onClick}>
      {children}
    </button>
  )
}

export default Button
