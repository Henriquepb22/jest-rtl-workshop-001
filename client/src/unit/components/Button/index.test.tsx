import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Button from './index'

describe('<Button />', () => {
  it('should render component', () => {
    const { container } = render(<Button />)

    expect(container).toMatchSnapshot()
  })

  it('should render the medium size be default', () => {
    render(<Button />)

    expect(screen.getByRole('button')).toHaveClass('px-4 py-2 text-base')
  })

  it('should render the primary color be default', () => {
    render(<Button />)

    expect(screen.getByRole('button')).toHaveClass('bg-pink-600 text-white')
  })

  it('should render the secondary color', () => {
    render(<Button color="secondary" />)

    expect(screen.getByRole('button')).toHaveClass('bg-gray-300 text-black')
  })

  it('should render the small button variant', () => {
    render(<Button size="sm" />)

    expect(screen.getByRole('button')).toHaveClass('px-2 py-1 text-sm')
  })

  it('should render the big button variant', () => {
    render(<Button size="lg" />)

    expect(screen.getByRole('button')).toHaveClass('px-6 py-3 text-lg')
  })

  it('should render a disabled button', () => {
    render(<Button disabled />)

    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('should render a hover opacity', () => {
    render(<Button disabled />)

    userEvent.hover(screen.getByRole('button'))
    expect(screen.getByRole('button')).toHaveClass('hover:opacity-80')
  })
})
