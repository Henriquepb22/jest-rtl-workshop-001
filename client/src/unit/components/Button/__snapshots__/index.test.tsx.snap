// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render component 1`] = `
<div>
  <button
    class="px-4 py-2 text-base bg-pink-600 text-white rounded w-full transition hover:opacity-80 disabled:opacity-70 disabled:cursor-not-allowed"
  />
</div>
`;
